modules/message-lane/src/lib.rs

```diff
@@ -262,13 +278,31 @@ decl_module! {
 				return Err(Error::<T, I>::InvalidMessagesDispatchWeight.into());
 			}
 
-			// dispatch messages
-			let total_messages = messages.len();
+			// dispatch messages and (optionally) update lane(s) state(s)
+			let mut total_messages = 0;
 			let mut valid_messages = 0;
-			for message in messages {
-				let mut lane = inbound_lane::<T, I>(message.key.lane_id);
-				if lane.receive_message::<T::MessageDispatch>(message.key.nonce, message.data) {
-					valid_messages += 1;
+			for (lane_id, lane_data) in messages {
+				let mut lane = inbound_lane::<T, I>(lane_id);
+
+				if let Some(lane_state) = lane_data.lane_state {
+					let updated_latest_confirmed_nonce = lane.receive_state_update(lane_state);
+					if let Some(updated_latest_confirmed_nonce) = updated_latest_confirmed_nonce {
+						frame_support::debug::trace!(
+							target: "runtime",
+							"Received lane {:?} state update: latest_confirmed_nonce={}",
+							lane_id,
+							updated_latest_confirmed_nonce,
+						);
+					}
+				}
+
+				for message in lane_data.messages {
+					debug_assert_eq!(message.key.lane_id, lane_id);
+
+					total_messages += 1;
+					if lane.receive_message::<T::MessageDispatch>(relayer_id.clone(), message.key.nonce, message.data) {
+						valid_messages += 1;
```

Comment:
Why don't we reject the entire transaction here? Previously we would get `false` here in case there was a nonce mismatch, now it may also happen if there is simply too many unconfirmed deliveries. I feel this is not something that is super desirable, but at the same time can't find any significant problems with the current approach.

Reply:
I was initially assuming that the delivery proof is going to be simply the transaction proof, but I've realised is that this makes the `verify_messages_delivery_proof` function harder to implement and potentially increase the size of the proof unnecessarily.
It seems that you assume that the delivery proof is actually going to be simply a storage proof, we don't necessarily care that much.

Reply:
> Why don't we reject the entire transaction here? Previously we would get false here in case there was a nonce mismatch, now it may also happen if there is simply too many unconfirmed deliveries. I feel this is not something that is super desirable, but at the same time can't find any significant problems with the current approach.

Because, as it was previously, transactions may overlap. Tx1[msg1, msg2, msg3] from relayer1 may be mined earler than Tx2[msg1, msg2, msg3, msg4] from relayer2, but it still may bring useful data (msg4). Relayer has all means to determine when it need to include `proof-of(outbound_lane.state)` into delivery transaction.